File: backend/app/teaching/services/staff.py
"""Staff service: business logic; uses repository for DB."""
import calendar
from datetime import date
from decimal import Decimal
from uuid import UUID

from sqlalchemy import select, and_
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.exceptions import ConflictError, NotFoundError
from app.teaching.models.staff import Staff, SalaryPayment
from app.teaching.models.leave import LeaveRequest
from app.teaching.schemas.staff import (
    StaffCreate,
    StaffUpdate,
    SalaryPaymentCreate,
    SalaryPaymentUpdate,
    BulkSalaryPaymentItem,
    LeaveSummaryResponse,
    TransferStaffCreate,
)
from app.teaching.repositories.staff import staff_repository


def _get_days_in_month(year: int, month: int) -> int:
    """Return number of days in a given month."""
    return calendar.monthrange(year, month)[1]


def _calculate_salary_breakdown(
    monthly_salary: Decimal,
    per_day_salary: Decimal | None,
    allowed_leaves: int,
    leaves_taken: int,
    days_worked: int,
    extra_allowance: Decimal,
    extra_deduction: Decimal,
) -> tuple[Decimal, Decimal, int, Decimal]:
    """
    Calculate salary breakdown.
    Returns: (per_day_rate, leave_deduction, excess_leaves, calculated_salary)
    """
    per_day = per_day_salary if per_day_salary else (monthly_salary / Decimal("30"))
    excess_leaves = max(0, leaves_taken - allowed_leaves)
    leave_deduction = Decimal(str(excess_leaves)) * per_day
    calculated_salary = monthly_salary - leave_deduction - extra_deduction + extra_allowance
    return per_day, leave_deduction, excess_leaves, calculated_salary


class StaffService:
    async def create(self, db: AsyncSession, data: StaffCreate) -> Staff:
        staff = Staff(**data.model_dump())
        return await staff_repository.add(db, staff)

    async def create_many(self, db: AsyncSession, items: list[StaffCreate]) -> list[Staff]:
        out = []
        for d in items:
            staff = Staff(**d.model_dump())
            out.append(await staff_repository.add(db, staff))
        return out

    async def get(self, db: AsyncSession, id: UUID) -> Staff | None:
        return await staff_repository.get(db, id)

    async def get_or_404(self, db: AsyncSession, id: UUID) -> Staff:
        staff = await staff_repository.get(db, id)
        if not staff:
            raise NotFoundError("Staff not found")
        return staff

    async def delete_all_by_session(self, db: AsyncSession, session_id: UUID) -> int:
        """Delete all staff in the given session. Returns count deleted."""
        return await staff_repository.delete_all_by_session(db, session_id)

    async def list_by_session(self, db: AsyncSession, session_id: UUID) -> list[Staff]:
        return await staff_repository.list_by_session(db, session_id)

    async def list_all(self, db: AsyncSession) -> list[Staff]:
        return await staff_repository.list_all(db)

    async def list_all_paginated(
        self,
        db: AsyncSession,
        *,
        limit: int,
        offset: int,
    ) -> tuple[list[Staff], int]:
        """Return (items, total) for DB-level pagination when no session/org filter."""
        total = await staff_repository.count_all(db)
        items = await staff_repository.list_all_paginated(
            db, limit=limit, offset=offset
        )
        return items, total

    async def list_by_organization(self, db: AsyncSession, organization_id: UUID) -> list[Staff]:
        return await staff_repository.list_by_organization(db, organization_id)

    async def list_by_session_paginated(
        self,
        db: AsyncSession,
        session_id: UUID,
        *,
        limit: int,
        offset: int,
        search: str | None = None,
        role: str | None = None,
    ) -> tuple[list[Staff], int]:
        total = await staff_repository.count_by_session(
            db, session_id, search=search, role=role
        )
        items = await staff_repository.list_by_session_paginated(
            db,
            session_id,
            limit=limit,
            offset=offset,
            search=search,
            role=role,
        )
        return items, total

    async def transfer_to_session(
        self, db: AsyncSession, data: TransferStaffCreate
    ) -> int:
        """Copy staff from one session to another. Copies salary and other details; does not copy salary payment records."""
        from_session_id = data.from_session_id
        to_session_id = data.to_session_id
        count = 0
        for staff_id in data.staff_ids:
            staff = await staff_repository.get(db, staff_id)
            if not staff or staff.session_id != from_session_id:
                continue
            new_staff = Staff(
                session_id=to_session_id,
                user_id=None,
                name=staff.name,
                employee_id=staff.employee_id,
                role=staff.role,
                monthly_salary=staff.monthly_salary,
                subject_or_grade=staff.subject_or_grade,
                phone=staff.phone,
                email=staff.email,
                address=staff.address,
                salary_due_day=staff.salary_due_day,
                allowed_leaves_per_month=staff.allowed_leaves_per_month,
                per_day_salary=staff.per_day_salary,
                classes_subjects=staff.classes_subjects,
            )
            await staff_repository.add(db, new_staff)
            count += 1
        return count

    async def list_by_organization_paginated(
        self,
        db: AsyncSession,
        organization_id: UUID,
        *,
        limit: int,
        offset: int,
    ) -> tuple[list[Staff], int]:
        total = await staff_repository.count_by_organization(db, organization_id)
        items = await staff_repository.list_by_organization_paginated(
            db, organization_id, limit=limit, offset=offset
        )
        return items, total

    async def update(self, db: AsyncSession, id: UUID, data: StaffUpdate) -> Staff:
        staff = await self.get_or_404(db, id)
        for k, v in data.model_dump(exclude_unset=True).items():
            setattr(staff, k, v)
        return await staff_repository.update(db, staff)

    async def delete(self, db: AsyncSession, id: UUID) -> None:
        staff = await self.get_or_404(db, id)
        await staff_repository.delete(db, staff)

    async def get_leave_summary(
        self, db: AsyncSession, staff_id: UUID, month: str
    ) -> LeaveSummaryResponse:
        """
        Get leave summary for a staff member for a specific month.
        Returns leaves taken (approved), days in month, calculated days worked, etc.
        """
        staff = await self.get_or_404(db, staff_id)
        
        # Parse month (YYYY-MM)
        year, month_num = map(int, month.split("-"))
        days_in_month = _get_days_in_month(year, month_num)
        
        # Get first and last day of the month
        first_day = date(year, month_num, 1)
        last_day = date(year, month_num, days_in_month)
        
        # Query approved leave requests for this staff in this month
        result = await db.execute(
            select(LeaveRequest).where(
                and_(
                    LeaveRequest.staff_id == staff_id,
                    LeaveRequest.status == "approved",
                    LeaveRequest.from_date <= last_day,
                    LeaveRequest.to_date >= first_day,
                )
            )
        )
        leave_requests = list(result.scalars().all())
        
        # Calculate total leave days in this month
        leaves_taken = 0
        for req in leave_requests:
            # Calculate overlapping days with this month
            overlap_start = max(req.from_date, first_day)
            overlap_end = min(req.to_date, last_day)
            if overlap_end >= overlap_start:
                leaves_taken += (overlap_end - overlap_start).days + 1
        
        allowed_leaves = staff.allowed_leaves_per_month
        excess_leaves = max(0, leaves_taken - allowed_leaves)
        days_worked = days_in_month - leaves_taken
        
        per_day = staff.per_day_salary if staff.per_day_salary else (staff.monthly_salary / Decimal("30"))
        leave_deduction = Decimal(str(excess_leaves)) * per_day
        
        return LeaveSummaryResponse(
            staff_id=staff_id,
            month=month,
            leaves_taken=leaves_taken,
            days_in_month=days_in_month,
            days_worked=days_worked,
            allowed_leaves=allowed_leaves,
            excess_leaves=excess_leaves,
            per_day_salary=per_day,
            leave_deduction=leave_deduction,
        )

    async def add_salary_payment(
        self, db: AsyncSession, staff_id: UUID, data: SalaryPaymentCreate
    ) -> SalaryPayment:
        staff = await self.get_or_404(db, staff_id)

        # Check if there's an existing payment for this month
        existing_payment = await staff_repository.get_salary_payment_for_month(db, staff_id, data.month)
        
        # Calculate salary breakdown
        per_day, leave_deduction, excess_leaves, calculated_salary = _calculate_salary_breakdown(
            monthly_salary=staff.monthly_salary,
            per_day_salary=staff.per_day_salary,
            allowed_leaves=staff.allowed_leaves_per_month,
            leaves_taken=data.leaves_taken,
            days_worked=data.days_worked,
            extra_allowance=data.extra_allowance,
            extra_deduction=data.extra_deduction,
        )
        
        # Use provided paid_amount or fall back to expected amount
        paid_amount = data.paid_amount if data.paid_amount is not None else data.amount
        
        if existing_payment:
            # Update existing payment for partial payment scenario
            existing_payment.paid_amount = paid_amount
            existing_payment.status = data.status
            existing_payment.payment_date = data.payment_date
            existing_payment.method = data.method
            return await staff_repository.update(db, existing_payment)
        
        return await staff_repository.add_salary_payment(
            db,
            staff_id,
            month=data.month,
            amount=data.amount,
            paid_amount=paid_amount,
            status=data.status,
            payment_date=data.payment_date,
            method=data.method,
            due_date=data.due_date,
            # Leave tracking
            days_worked=data.days_worked,
            leaves_taken=data.leaves_taken,
            allowed_leaves=staff.allowed_leaves_per_month,
            excess_leaves=excess_leaves,
            leave_deduction=leave_deduction,
            # Extra allowance/deduction
            extra_allowance=data.extra_allowance,
            allowance_note=data.allowance_note,
            extra_deduction=data.extra_deduction,
            deduction_note=data.deduction_note,
            # Calculated salary
            calculated_salary=calculated_salary,
        )

    async def update_salary_payment(
        self, db: AsyncSession, staff_id: UUID, payment_id: UUID, data: SalaryPaymentUpdate
    ) -> SalaryPayment | None:
        await self.get_or_404(db, staff_id)
        return await staff_repository.update_salary_payment(
            db,
            payment_id,
            staff_id,
            paid_amount=data.paid_amount,
            status=data.status,
            payment_date=data.payment_date,
            method=data.method,
            days_worked=data.days_worked,
            leaves_taken=data.leaves_taken,
            extra_allowance=data.extra_allowance,
            allowance_note=data.allowance_note,
            extra_deduction=data.extra_deduction,
            deduction_note=data.deduction_note,
        )

    async def add_salary_payments_bulk(
        self, db: AsyncSession, items: list[BulkSalaryPaymentItem]
    ) -> list[SalaryPayment]:
        # Pre-fetch all staff members
        staff_map: dict[UUID, Staff] = {}
        for sid in {item.staff_id for item in items}:
            staff = await self.get_or_404(db, sid)
            staff_map[sid] = staff

        # Reject if any employee already has a payment for that month
        for item in items:
            if await staff_repository.has_salary_payment_for_month(db, item.staff_id, item.month):
                staff = staff_map[item.staff_id]
                raise ConflictError(
                    f"Employee {staff.name} already has a salary payment for {item.month}. Only one payment per employee per month is allowed."
                )

        out = []
        for item in items:
            staff = staff_map[item.staff_id]

            # Calculate salary breakdown
            per_day, leave_deduction, excess_leaves, calculated_salary = _calculate_salary_breakdown(
                monthly_salary=staff.monthly_salary,
                per_day_salary=staff.per_day_salary,
                allowed_leaves=staff.allowed_leaves_per_month,
                leaves_taken=item.leaves_taken,
                days_worked=item.days_worked,
                extra_allowance=item.extra_allowance,
                extra_deduction=item.extra_deduction,
            )
            
            payment = await staff_repository.add_salary_payment(
                db,
                item.staff_id,
                month=item.month,
                amount=item.amount,
                status=item.status,
                payment_date=item.payment_date,
                method=item.method,
                due_date=item.due_date,
                # Leave tracking
                days_worked=item.days_worked,
                leaves_taken=item.leaves_taken,
                allowed_leaves=staff.allowed_leaves_per_month,
                excess_leaves=excess_leaves,
                leave_deduction=leave_deduction,
                # Extra allowance/deduction
                extra_allowance=item.extra_allowance,
                allowance_note=item.allowance_note,
                extra_deduction=item.extra_deduction,
                deduction_note=item.deduction_note,
                # Calculated salary
                calculated_salary=calculated_salary,
            )
            out.append(payment)
        return out

    async def delete_salary_payment(
        self, db: AsyncSession, staff_id: UUID, payment_id: UUID
    ) -> bool:
        await self.get_or_404(db, staff_id)
        return await staff_repository.delete_salary_payment(db, payment_id, staff_id)


staff_service = StaffService()

File: school/src/lib/db/repositories/staff.ts
import { getSupabase } from '../supabase';
import type { Staff, SalaryPayment, PaymentMethod, ClassSubject } from '../../../types';
import type { PaginatedResult } from './schools';

type SalaryStatus = SalaryPayment['status'];

export interface ExtendedSalaryPayment extends SalaryPayment {
  expectedAmount?: number;
  lateDays?: number;
}

function mapClassesSubjects(raw: unknown): ClassSubject[] | undefined {
  if (!Array.isArray(raw)) return undefined;
  return raw.map((item: Record<string, unknown>) => ({
    className: String(item.class_name ?? item.className ?? ''),
    subjects: Array.isArray(item.subjects) ? item.subjects.map(String) : [],
  }));
}

function dbRowToStaff(row: Record<string, unknown>, salaryPayments: ExtendedSalaryPayment[] = []): Staff {
  return {
    id: row.id as string,
    sessionId: row.session_id as string,
    name: row.name as string,
    employeeId: row.employee_id as string,
    role: row.role as Staff['role'],
    monthlySalary: Number(row.monthly_salary) || 0,
    subjectOrGrade: row.subject_or_grade as string | undefined,
    // Leave & salary deduction configuration
    allowedLeavesPerMonth: Number(row.allowed_leaves_per_month) || 1,
    perDaySalary: row.per_day_salary != null ? Number(row.per_day_salary) : undefined,
    // Classes & subjects
    classesSubjects: mapClassesSubjects(row.classes_subjects),
    // Personal details
    aadhaarNumber: row.aadhaar_number as string | undefined,
    dateOfBirth: row.date_of_birth as string | undefined,
    salaryPayments,
  };
}

function mapSalaryPaymentFromDb(p: Record<string, unknown>): ExtendedSalaryPayment {
  return {
    id: p.id as string,
    month: p.month as string,
    amount: Number(p.paid_amount) || 0,
    paidAmount: Number(p.paid_amount) || 0,
    status: (p.status as string) as SalaryStatus,
    paymentDate: p.payment_date as string | undefined,
    method: (p.method as string | undefined) as PaymentMethod | undefined,
    expectedAmount: Number(p.expected_amount),
    dueDate: p.due_date as string | undefined,
    lateDays: p.late_days as number | undefined,
    // Leave tracking fields
    daysWorked: Number(p.days_worked) || 30,
    leavesTaken: Number(p.leaves_taken) || 0,
    allowedLeaves: Number(p.allowed_leaves) || 1,
    excessLeaves: Number(p.excess_leaves) || 0,
    leaveDeduction: Number(p.leave_deduction) || 0,
    // Extra allowance/deduction
    extraAllowance: Number(p.extra_allowance) || 0,
    allowanceNote: p.allowance_note as string | undefined,
    extraDeduction: Number(p.extra_deduction) || 0,
    deductionNote: p.deduction_note as string | undefined,
    // Calculated salary
    calculatedSalary: Number(p.calculated_salary) || Number(p.paid_amount) || 0,
  };
}

export const staffRepository = {
  async getAll(): Promise<Staff[]> {
    const supabase = getSupabase();
    const { data, error } = await supabase
      .from('school_xx_staff')
      .select('*')
      .order('name');

    if (error) throw new Error('Failed to fetch staff');

    const staffIds = (data || []).map((s: { id: string }) => s.id);
    const { data: salaryData } = await supabase
      .from('school_xx_salary_payments')
      .select('*')
      .in('staff_id', staffIds);

    const salaryByStaff: Record<string, ExtendedSalaryPayment[]> = {};
    (salaryData || []).forEach((p: Record<string, unknown>) => {
      const sid = p.staff_id as string;
      if (!salaryByStaff[sid]) salaryByStaff[sid] = [];
      salaryByStaff[sid].push(mapSalaryPaymentFromDb(p));
    });

    return (data || []).map((row: Record<string, unknown>) => dbRowToStaff(row, salaryByStaff[row.id as string] || []));
  },

  async getBySession(sessionId: string): Promise<Staff[]> {
    const supabase = getSupabase();
    const { data, error } = await supabase
      .from('school_xx_staff')
      .select('*')
      .eq('session_id', sessionId)
      .order('name');

    if (error) throw new Error('Failed to fetch staff');

    const staffIds = (data || []).map((s: { id: string }) => s.id);
    const { data: salaryData } = staffIds.length > 0
      ? await supabase.from('school_xx_salary_payments').select('*').in('staff_id', staffIds)
      : { data: [] };

    const salaryByStaff: Record<string, ExtendedSalaryPayment[]> = {};
    (salaryData || []).forEach((p: Record<string, unknown>) => {
      const sid = p.staff_id as string;
      if (!salaryByStaff[sid]) salaryByStaff[sid] = [];
      salaryByStaff[sid].push(mapSalaryPaymentFromDb(p));
    });

    return (data || []).map((row: Record<string, unknown>) => dbRowToStaff(row, salaryByStaff[row.id as string] || []));
  },

  async getPaginated(
    page: number = 1,
    pageSize: number = 10,
    filters?: { sessionId?: string; role?: string; search?: string }
  ): Promise<PaginatedResult<Staff>> {
    const supabase = getSupabase();
    const offset = (page - 1) * pageSize;

    let query = supabase
      .from('school_xx_staff')
      .select('*', { count: 'exact' })
      .order('name');

    if (filters?.sessionId) query = query.eq('session_id', filters.sessionId);
    if (filters?.role) query = query.eq('role', filters.role);
    if (filters?.search) query = query.or(`name.ilike.%${filters.search}%,employee_id.ilike.%${filters.search}%`);

    const { data, error, count } = await query.range(offset, offset + pageSize - 1);

    if (error) throw new Error('Failed to fetch staff');

    const staffIds = (data || []).map((s: { id: string }) => s.id);
    const { data: salaryData } = staffIds.length > 0
      ? await supabase.from('school_xx_salary_payments').select('*').in('staff_id', staffIds)
      : { data: [] };

    const salaryByStaff: Record<string, ExtendedSalaryPayment[]> = {};
    (salaryData || []).forEach((p: Record<string, unknown>) => {
      const sid = p.staff_id as string;
      if (!salaryByStaff[sid]) salaryByStaff[sid] = [];
      salaryByStaff[sid].push(mapSalaryPaymentFromDb(p));
    });

    const staff = (data || []).map((row: Record<string, unknown>) => dbRowToStaff(row, salaryByStaff[row.id as string] || []));

    return {
      data: staff,
      total: count || 0,
      page,
      pageSize,
      totalPages: Math.ceil((count || 0) / pageSize),
    };
  },

  async getById(id: string): Promise<Staff | null> {
    const supabase = getSupabase();
    const { data, error } = await supabase
      .from('school_xx_staff')
      .select('*')
      .eq('id', id)
      .single();

    if (error || !data) return null;

    const { data: salaryData } = await supabase
      .from('school_xx_salary_payments')
      .select('*')
      .eq('staff_id', id);

    const salaryPayments: ExtendedSalaryPayment[] = (salaryData || []).map(mapSalaryPaymentFromDb);

    return dbRowToStaff(data, salaryPayments);
  },

  async create(staffMember: Omit<Staff, 'id' | 'salaryPayments'>): Promise<Staff> {
    const supabase = getSupabase();
    const id = crypto.randomUUID();

    const row: Record<string, unknown> = {
      id,
      session_id: staffMember.sessionId,
      name: staffMember.name,
      employee_id: staffMember.employeeId,
      role: staffMember.role,
      monthly_salary: staffMember.monthlySalary,
      subject_or_grade: staffMember.subjectOrGrade ?? null,
      allowed_leaves_per_month: staffMember.allowedLeavesPerMonth ?? 1,
      per_day_salary: staffMember.perDaySalary ?? null,
    };
    if (staffMember.classesSubjects?.length) {
      row.classes_subjects = staffMember.classesSubjects.map((cs) => ({
        class_name: cs.className,
        subjects: cs.subjects,
      }));
    }

    const { data, error } = await supabase
      .from('school_xx_staff')
      .insert(row)
      .select()
      .single();

    if (error) throw new Error(error.message || 'Failed to create staff member');
    return dbRowToStaff(data, []);
  },

  async createMany(staffMembers: Omit<Staff, 'id' | 'salaryPayments'>[]): Promise<Staff[]> {
    if (staffMembers.length === 0) return [];
    const supabase = getSupabase();
    const rows = staffMembers.map((s) => {
      const row: Record<string, unknown> = {
        id: crypto.randomUUID(),
        session_id: s.sessionId,
        name: s.name,
        employee_id: s.employeeId,
        role: s.role,
        monthly_salary: s.monthlySalary,
        subject_or_grade: s.subjectOrGrade ?? null,
        allowed_leaves_per_month: s.allowedLeavesPerMonth ?? 1,
        per_day_salary: s.perDaySalary ?? null,
      };
      if (s.classesSubjects?.length) {
        row.classes_subjects = s.classesSubjects.map((cs) => ({
          class_name: cs.className,
          subjects: cs.subjects,
        }));
      }
      return row;
    });
    const { data, error } = await supabase.from('school_xx_staff').insert(rows).select();
    if (error) throw new Error(error.message || 'Failed to create staff members');
    return (data || []).map((row: Record<string, unknown>) => dbRowToStaff(row, []));
  },

  async update(id: string, updates: Partial<Omit<Staff, 'id' | 'salaryPayments'>>): Promise<Staff> {
    const supabase = getSupabase();
    const dbUpdates: Record<string, unknown> = {};
    if (updates.sessionId !== undefined) dbUpdates.session_id = updates.sessionId;
    if (updates.name !== undefined) dbUpdates.name = updates.name;
    if (updates.employeeId !== undefined) dbUpdates.employee_id = updates.employeeId;
    if (updates.role !== undefined) dbUpdates.role = updates.role;
    if (updates.monthlySalary !== undefined) dbUpdates.monthly_salary = updates.monthlySalary;
    if (updates.subjectOrGrade !== undefined) dbUpdates.subject_or_grade = updates.subjectOrGrade;

    const { data, error } = await supabase
      .from('school_xx_staff')
      .update(dbUpdates)
      .eq('id', id)
      .select()
      .single();

    if (error) throw new Error('Failed to update staff member');

    const { data: salaryData } = await supabase
      .from('school_xx_salary_payments')
      .select('*')
      .eq('staff_id', id);

    const salaryPayments: ExtendedSalaryPayment[] = (salaryData || []).map(mapSalaryPaymentFromDb);

    return dbRowToStaff(data, salaryPayments);
  },

  async delete(id: string): Promise<void> {
    const supabase = getSupabase();
    const { error } = await supabase
      .from('school_xx_staff')
      .delete()
      .eq('id', id);

    if (error) throw new Error('Failed to delete staff member');
  },

  /** Delete all staff in the given session. Returns number deleted. */
  async deleteAllBySession(sessionId: string): Promise<number> {
    const supabase = getSupabase();
    const { data, error } = await supabase
      .from('school_xx_staff')
      .delete()
      .eq('session_id', sessionId)
      .select('id');
    if (error) throw new Error('Failed to delete all staff');
    return data?.length ?? 0;
  },

  /** Copy staff to another session; salary payment records are not copied. */
  async transferToSession(params: {
    fromSessionId: string;
    staffIds: string[];
    toSessionId: string;
  }): Promise<number> {
    const { staffIds, toSessionId } = params;
    if (staffIds.length === 0) return 0;
    const supabase = getSupabase();
    const { data: rows, error: fetchError } = await supabase
      .from('school_xx_staff')
      .select('*')
      .in('id', staffIds)
      .eq('session_id', params.fromSessionId);
    if (fetchError || !rows?.length) return 0;
    const inserts = rows.map((row: Record<string, unknown>) => ({
      id: crypto.randomUUID(),
      session_id: toSessionId,
      name: row.name,
      employee_id: row.employee_id,
      role: row.role,
      monthly_salary: row.monthly_salary,
      subject_or_grade: row.subject_or_grade ?? null,
      phone: row.phone ?? null,
      email: row.email ?? null,
      address: row.address ?? null,
      salary_due_day: row.salary_due_day ?? 5,
      allowed_leaves_per_month: row.allowed_leaves_per_month ?? 2,
      per_day_salary: row.per_day_salary ?? null,
      classes_subjects: row.classes_subjects ?? null,
    }));
    const { error: insertError } = await supabase.from('school_xx_staff').insert(inserts);
    if (insertError) throw new Error('Failed to transfer staff');
    return inserts.length;
  },

  async addSalaryPayment(staffId: string, payment: Omit<ExtendedSalaryPayment, 'id' | 'lateDays'> & { paidAmount?: number }): Promise<ExtendedSalaryPayment> {
    const supabase = getSupabase();
    const dueDate = payment.dueDate || `${payment.month}-05`;
    
    // Check if a payment already exists for this month (for partial payment continuation)
    const { data: existing } = await supabase
      .from('school_xx_salary_payments')
      .select('id')
      .eq('staff_id', staffId)
      .eq('month', payment.month)
      .single();

    const paidAmount = payment.paidAmount ?? payment.amount;
    const calculatedSalary = payment.calculatedSalary ?? payment.amount;

    const paymentData = {
      staff_id: staffId,
      month: payment.month,
      expected_amount: calculatedSalary,
      paid_amount: paidAmount,
      status: payment.status,
      due_date: dueDate,
      payment_date: payment.paymentDate,
      method: payment.method,
      days_worked: payment.daysWorked ?? 30,
      leaves_taken: payment.leavesTaken ?? 0,
      allowed_leaves: payment.allowedLeaves ?? 1,
      excess_leaves: payment.excessLeaves ?? 0,
      leave_deduction: payment.leaveDeduction ?? 0,
      extra_allowance: payment.extraAllowance ?? 0,
      allowance_note: payment.allowanceNote,
      extra_deduction: payment.extraDeduction ?? 0,
      deduction_note: payment.deductionNote,
      calculated_salary: calculatedSalary,
    };

    let data;
    let error;

    if (existing) {
      // Update existing payment (partial payment scenario)
      const result = await supabase
        .from('school_xx_salary_payments')
        .update(paymentData)
        .eq('id', existing.id)
        .select()
        .single();
      data = result.data;
      error = result.error;
    } else {
      // Insert new payment
      const result = await supabase
        .from('school_xx_salary_payments')
        .insert({ id: crypto.randomUUID(), ...paymentData })
        .select()
        .single();
      data = result.data;
      error = result.error;
    }

    if (error) throw new Error('Failed to add salary payment');

    return {
      id: data.id,
      month: data.month,
      amount: Number(data.paid_amount),
      paidAmount: Number(data.paid_amount),
      status: data.status as SalaryStatus,
      paymentDate: data.payment_date,
      method: data.method as PaymentMethod | undefined,
      expectedAmount: Number(data.expected_amount),
      dueDate: data.due_date,
      lateDays: data.late_days,
      daysWorked: Number(data.days_worked) || 30,
      leavesTaken: Number(data.leaves_taken) || 0,
      allowedLeaves: Number(data.allowed_leaves) || 1,
      excessLeaves: Number(data.excess_leaves) || 0,
      leaveDeduction: Number(data.leave_deduction) || 0,
      extraAllowance: Number(data.extra_allowance) || 0,
      allowanceNote: data.allowance_note,
      extraDeduction: Number(data.extra_deduction) || 0,
      deductionNote: data.deduction_note,
      calculatedSalary: Number(data.calculated_salary) || Number(data.paid_amount) || 0,
    };
  },

  async addSalaryPaymentsBatch(
    payments: { staffId: string; payment: Omit<ExtendedSalaryPayment, 'id' | 'lateDays'> }[]
  ): Promise<void> {
    if (payments.length === 0) return;
    const supabase = getSupabase();
    const rows = payments.map(({ staffId, payment }) => ({
      id: crypto.randomUUID(),
      staff_id: staffId,
      month: payment.month,
      expected_amount: payment.expectedAmount ?? payment.amount,
      paid_amount: payment.amount,
      status: payment.status,
      due_date: payment.dueDate || `${payment.month}-05`,
      payment_date: payment.paymentDate,
      method: payment.method,
    }));
    const { error } = await supabase.from('school_xx_salary_payments').insert(rows);
    if (error) throw new Error('Failed to add salary payments');
  },

  async updateSalaryPayment(_staffId: string, paymentId: string, updates: Partial<ExtendedSalaryPayment>): Promise<ExtendedSalaryPayment> {
    const supabase = getSupabase();
    const dbUpdates: Record<string, unknown> = {};
    if (updates.month !== undefined) dbUpdates.month = updates.month;
    if (updates.amount !== undefined) dbUpdates.paid_amount = updates.amount;
    if (updates.expectedAmount !== undefined) dbUpdates.expected_amount = updates.expectedAmount;
    if (updates.status !== undefined) dbUpdates.status = updates.status;
    if (updates.paymentDate !== undefined) dbUpdates.payment_date = updates.paymentDate;
    if (updates.method !== undefined) dbUpdates.method = updates.method;
    if (updates.dueDate !== undefined) dbUpdates.due_date = updates.dueDate;

    const { data, error } = await supabase
      .from('school_xx_salary_payments')
      .update(dbUpdates)
      .eq('id', paymentId)
      .select()
      .single();

    if (error) throw new Error('Failed to update salary payment');

    return {
      id: data.id,
      month: data.month,
      amount: Number(data.paid_amount),
      status: data.status as SalaryStatus,
      paymentDate: data.payment_date,
      method: data.method as PaymentMethod | undefined,
      expectedAmount: Number(data.expected_amount),
      dueDate: data.due_date,
      lateDays: data.late_days,
      daysWorked: Number(data.days_worked) || 30,
      leavesTaken: Number(data.leaves_taken) || 0,
      allowedLeaves: Number(data.allowed_leaves) || 2,
      excessLeaves: Number(data.excess_leaves) || 0,
      leaveDeduction: Number(data.leave_deduction) || 0,
      extraAllowance: Number(data.extra_allowance) || 0,
      extraDeduction: Number(data.extra_deduction) || 0,
      calculatedSalary: Number(data.calculated_salary) || Number(data.paid_amount) || 0,
    };
  },

  async deleteSalaryPayment(_staffId: string, paymentId: string): Promise<void> {
    const supabase = getSupabase();
    const { error } = await supabase
      .from('school_xx_salary_payments')
      .delete()
      .eq('id', paymentId);

    if (error) throw new Error('Failed to delete salary payment');
  },

  async getLastPaidDate(staffId: string): Promise<string | null> {
    const staff = await this.getById(staffId);
    if (!staff) return null;

    const paidPayments = staff.salaryPayments
      .filter(p => p.status === 'Paid' && p.paymentDate)
      .sort((a, b) => new Date(b.paymentDate!).getTime() - new Date(a.paymentDate!).getTime());

    return paidPayments[0]?.paymentDate || null;
  },

  async getSalaryStatus(staffId: string, month: string): Promise<ExtendedSalaryPayment | null> {
    const staff = await this.getById(staffId);
    if (!staff) return null;

    return (staff.salaryPayments.find(p => p.month === month) as ExtendedSalaryPayment) || null;
  },
};
